Rewrite chunk system

Spottedleaf opened the pull request:
In the future, I want to experiment with ticking separate independent regions of the world in parallel. However, the current chunk system is just not up to that task: The scheduling is tied to the main thread, the chunk holder state is tied to the main thread, and the chunk system cannot fully utilise the worker threads it is given.

The new chunk system can be scheduled by any thread and as a result, the only time the chunk generation logic needs to go to the main thread is when it needs to execute generation/load logic that can only be done by the main thread (i.e loading in entity/poi chunks). As a result, chunk generation should be far less affected by lower TPS.

The new chunk system is also capable of executing several chunk status generation tasks in parallel. In my testing, chunky was capable of pushing all 8 threads I gave it for generating a new world, unlike the current chunk system.

Current things left to do before this can be merged:
- [x] Verify stability (i.e no save/unload/shutdown/crash issues)
- [x] Plugin compatibility
- [x] Figure out a merge strategy that will allow for this patch to be dropped for game updates, ensuring that it cannot stall any update
- [x] Complete the todo list

Current todo list:
- [x] Make the WatchdogThread and the main thread instances of TickThread and rewrite all of the main thread checks to just check for instanceof. This will result in allowing the watchdog thread to perform chunk system operations without worrying about stalling forever, which is a very common issue that prevents the world from properly saving.
- [x] Do not use ConcurrentHashMap for chunk holders. Need a solution that does not use any locks. Currently, CHM is used so the testing can validate the chunk system without worrying about the internal chunk holder map being broken.
- ~~Make player chunk scheduling asynchronous. Currently, the main thread has to aggressively add and propagate ticket levels and do the initial scheduling work, which isn't negligible and requires mostly good TPS. Once the chunk scheduling is moved to its own thread, chunk generation speed should be mostly independent of TPS and frees the main thread of the burden.~~
- [x] ALL of the TODO comments...

There are currently many aspects of the chunk system that need to be tested:
- Chunk generation
- Chunk loading
- Chunk (poi/entity/block) saving
- Chunk (poi/entity/block) unloading, **_especially_** with players that are passengers of an entity or entities that are passengers or vehicles
- All of the above combined with the server shutting down (there is a new shutdown process)
- Plugin compatibility (custom world generators for example)

In order to assist with debugging, I've rewritten `/paper debug chunks` and added a new command `/paper holderinfo`. The new paper holderinfo command can be used to see how many read only chunks are in memory as well as protochunks.

At this point, the chunk system is very new (in alpha) and as such will likely have problems that can result in data loss or even possibly corruption. Do not use this branch for anything you aren't OK with losing.

Current build: https://cdn.discordapp.com/attachments/876902758366736457/1020824757953904701/paper-paperclip-1.19.2-R0.1-SNAPSHOT-reobf.jar

Spottedleaf committed:
Rewrite chunk system initial commit

Spottedleaf committed:
Fix chunk full status updates being dropped during unload tick

Also ensure that during the unload tick that scheduling occurs
only out of the schedule lock.

Add more scheduling debug info to debug json and chunk task toString

Spottedleaf committed:
Make blocking poi/entity load set priority to URGENT

Spottedleaf committed:
Make shutdown save execute async unload task

When the server shuts down, the chunk system is halted.
So, any queued async chunk unload would not complete.
To fix this, just execute the async chunk unload task
synchronously in the save method.

Spottedleaf committed:
Stagger chunk unloading accross ticks

Unload at least 50 chunks or 5% of the unload queue,
whichever is larger

Spottedleaf committed:
Initialise priority correctly for I/O load task

Since it was not initialised, it would always be BLOCKING.

Spottedleaf committed:
Review todos

Spottedleaf committed:
Make server thread and watchdog thread instances of TickThread

Also change all thread checks to check for whether the
current thread is an instance of TickThread. This will
improve reliability on watchdog shutdown as now the
watchdog thread should now pass all of the thread checks
in the chunk system.

Spottedleaf committed:
Merge branch 'master' of github.com:PaperMC/Paper into dev/chunksystem

Spottedleaf committed:
Update to 1.19.1

Spottedleaf committed:
Merge branch 'master' of github.com:PaperMC/Paper into dev/chunksystem

Spottedleaf committed:
Re-add incremental chunk saving

Spottedleaf committed:
Make sure playerChunk field is initialised in LevelChunk

Fixes missing spawn/chunk ticks

masmc05 commented on patches/server/0924-Rewrite-chunk-system.patch:
```diff
@@ -0,0 +1,19703 @@
+     public static final class Holder {
+ 
+-        public static final PaperFileIOThread INSTANCE = new PaperFileIOThread();
++        public static final PaperFileIOThread INSTANCE = null; // Paper - fail hard on usage
```
so this is why dynmap fails to work

Is this temporary, or plugins will be required to use an alternative?

Spottedleaf replied:
I can add back in code to keep the api hooks here for 1.19, but I wont keep them for the next release (1.20)

masmc05 replied:
If they're breaking changes then better not, i'll be able to change the dynmap to test the speed, since before it wasn't able to reach full potential of my cpu on low quality render because was waiting for raw nbt chunk data from the io thread

Spottedleaf committed:
Add things into todo

Spottedleaf committed:
Merge branch 'master' of github.com:PaperMC/Paper into dev/chunksystem

Spottedleaf committed:
Rebuild patches

Spottedleaf committed:
Re-add async teleport

Spottedleaf committed:
Only set chunk status for protochunk

Spottedleaf committed:
Add support for reading chunk data with old regionfile API

deprecated and to be removed in 1.20

Spottedleaf committed:
Add unload queue to debuginfo and unload count to holderinfo

Spottedleaf committed:
Re-add watchdog dumps

Also make the final watchdog dump perform a chunk debug if
we were waiting on chunks

In order to ensure the watchdog thread cannot stall on
lock acquisition, the scheduling and tick locks have been
changed to ReentrantLock so that we can use tryLock with
a time out.

Even if the watchdog thread fails to acquire the locks, it will
still try to perform a dump.

Spottedleaf committed:
Add SWMRLong2ObjectHashTable to replace CHM in chunk holder manager

ConcurrentHashMap had a couple of problems:
1. Synchronization was overkill. For a performance critical map,
 we don't want synchronization when we don't need it.
2. Allocation of objects for map lookup is wasteful.
3. The hashcode for the Long object is awful. It is defined as
   `(x >>> 32) ^ x`, and for our long chunk key will result in a hashcode
   of `x ^ z`. This will force a significant number of collisions. For
   example, just me being logged into one location at ~10VD results in
   2400 chunkholders. With the Long object hashing, only 64 bin
   values were created for a table size of 2^15, thus requiring
   around 37 chained elements per bin. For the hashing seen in fastutil,
   the number of bin elements was ~2200, at the same table length.
   It's just not comparable.

Spottedleaf committed:
Merge branch 'master' of github.com:PaperMC/Paper into dev/chunksystem

Spottedleaf committed:
Rebuild patches

Spottedleaf committed:
Make getChunk block for BORDER full status when input status is full

The new chunk system considers the FULL chunk gen status only
for internal use, whereas the BORDER full status is for all external
use (i.e plugin events, game logic).

Change ChunkDataController's task map to use ChunkCoordinate
instead of Long, since Long uses a hashcode algorithm that
produces poor results for chunk keys.

Spottedleaf committed:
Only load in entity chunk for FULL chunk status progression

Before, EntityLookup would load in the entity check if it did
not exist. However, this was unneccessary. Instead, we just create
a transient entity chunk now.

We also want to restrain the EntitiesLoadEvent to the chunk full
status, rather than having it possibly be called for
things such as Entity#move, addEntity, etc, as we know that
the chunk full status is a safe place to call it.

Spottedleaf committed:
Merge branch 'master' of github.com:PaperMC/Paper into dev/chunksystem

Spottedleaf committed:
Fix NPE in getChunkAt for full status chunk

Spottedleaf committed:
Resolve some todos

Add warning if entity fails to add to entitychunk
Re-add implementations for CraftWorld getChunkCount
and getTileEntityCount

Spottedleaf committed:
Add duplicate entity config support

Note: As of writing, this is broken in current paper master because
paper master only uses the config for world gen entities, which
is the exact opposite of what it should be doing.

Spottedleaf committed:
Split concurrentutil and chunk system into two patches

Spottedleaf committed:
Merge branch 'master' of github.com:PaperMC/Paper into dev/chunksystem

Spottedleaf committed:
Resolve patch merge conflicts

Spottedleaf committed:
Properly call poi load callback

Looks like it wasn't being called before. Now call it when
loading in the full chunk data or when sync loading the poi chunk

Spottedleaf committed:
Do not log warning for when an entity is removed by duplicate entity config

Spottedleaf committed:
Respect accessible visibility status

The new chunk system made any entity accessible whether or not
the status was accessible to ensure that when a chunk was
eventually loaded to plugins, the entities would be visible.
However, this caused issues with the dynamic game event listener,
as it has no logic to properly deal with unloaded chunks in the
move() method, as it expects the accessible hooks to call
add/remove while the chunk is still loaded.

Fortunately, these changes still ensure that when the chunk
load event is called, that the entities will be accessible.

Spottedleaf committed:
Complete GenericDataLoadTask properly if completeOnMainOffMain completes

The stage must be advanced to completed before onComplete can
be called

Spottedleaf committed:
Fix NPE in maskNonAccessible

Spottedleaf committed:
Fix improper chunk full status change queueing

The old method first drained the queue into another local queue,
and then began processing from the local queue. However, this
caused problems if a full status change recursively went to process
other full status, it would never complete any from the first
local queue.

Spottedleaf committed:
Fix incorrect chunk full determination for neighbours

It should be only after the chunk is BORDER, not before - this
would make TICKING run before its 1 radius neighbours were
loaded, for example. Now, when TICKING runs, its 1 radius neighbours
should be considered loaded (same for ENTITY_TICKING and its 2 radius).

Spottedleaf committed:
Fix upgradeGenTarget when generation task is cancelled

It can still re-schedule provided the generation task
has not yet completed.

Spottedleaf committed:
Fix NPE when downgrading a chunk from full status

The neighbours could be already unloaded at this stage

Spottedleaf committed:
Improve recursive full chunk status update logic

Block recursive ticket level updates, which would cause
recursive full chunk status updates, for entity chunk
status changes. This shouldn't affect plugin compatibility,
because this is the behavior in the old chunk system. Besides,
the logic is just not safe to run recursively.

For all full upgrade statuses, set the current full chunk status
before running the callbacks, so that any further recursive upgrade
will not double call any callbacks.

Spottedleaf committed:
Resolve TODOs

Spottedleaf committed:
Change ordering of entities load/unload events

Fire the entities load event after the chunk load event,
and fire the entities unload event before the chunk unload event.

This should fix problems with deadlock during the chunk full task,
as many plugins appear to assume the entities load event
is always after a chunk load event (which it never was for all cases,
but that doesn't matter). Now, plugins using getChunk shouldn't
deadlock the server anymore.

Also, add a new command: paper debug chunks --async
which when ran from console, will not schedule to the main thread.
This allows completely dead servers to produce a chunk debug.

Spottedleaf committed:
Prevent recursive full status update if downgrading status

If something were to make the chunk ticking/entity ticking
during the unload event, the recursive update would
update the status to ticking/entity ticking, and then after
the recursive update the original update would set the status
to INACCESSIBLE, without calling the not-ticking/entityticking
callbacks.

This also fixes double calling the downgrade statuses recursively,
since the next state is only set after the callback - so any recursive
call would double call the callback.

Spottedleaf committed:
Add UNKNOWN ticket for sync chunk loads

This will guarantee that the returned chunk is actually still
loaded, and that the returned chunk cannot unload until
the next chunk source tick.

Spottedleaf committed:
Clamp minimum ticket level expire time to 1 tick

Negative values are reserved for internal use. Also, values
smaller than 1 are pointless.

Spottedleaf committed:
Delay scheduled tasks from sync gen cancellation

The tasks shouldn't be processed while holding the scheduling lock

Spottedleaf committed:
Do not schedule I/O tasks for entity/poi data while holding schedule lock

The main reason tasks are not scheduled while holding
any locks is because of performance. Scheduling the tasks
is not always cheap, and if they are not cheap then they
are holding up a critical chunk system lock.

Spottedleaf committed:
Clamp minimum ticket level expire time to 0 ticks

The value 0 is actually used to indicate no expiration. So,
the first commit clamping to 1 caused non-expirable tickets
to expire immediately on next tick. Oops.

Spottedleaf committed:
Add UNKNOWN ticket for async chunk loads

This brings async loading in line with sync loading, so that
the chunk will stay loaded until the next chunk source tick.

Spottedleaf committed:
More aggressive entity ticking status update checks

Unfortunately, the current system for entity ticking statuses
cannot handle recursion at all. Currently ticket level
updates are blocked, teleporting the entity is blocked,
and removing/adding the entity to worlds is blocked.

Spottedleaf committed:
Merge branch 'master' of github.com:PaperMC/Paper into dev/chunksystem

Spottedleaf committed:
Rebuild patches

Spottedleaf committed:
Dump all players into chunk debug

Sometimes players may not actually be in a loaded world,
so make sure we dump them all.

Spottedleaf committed:
Fix full status queueing for BORDER

The ticket level and neighbour update hooks need to respect a pending
BORDER update. The ticket updates should only be after the
full BORDER queue.

Spottedleaf committed:
Merge branch 'master' of github.com:PaperMC/Paper into dev/chunksystem

Spottedleaf committed:
Rebuild patches

Spottedleaf committed:
Re-arrange most chunk system patches to front

Moved most chunk system hooks to a single class, so that
later patches are not hopefully reliant on specific
patches even exist.

Spottedleaf committed:
Fixes

- For scheduling a ticket state, make the callback accept
  LevelChunk instead of ChunkAccess
- Add POI_LOAD ticket for loading poi chunks
- Remove ENTITY_LOAD ticket if entity chunk becomes
  suddenly available after acquiring lock
- Remove removeDelay api from ticket levels, as
  replacement isn't well defined if removeDelay > 0. It's
  just better to use a different ticket entirely for a
  different delay.
- Only retain chunk data if it fails I/O when writing

Spottedleaf committed:
Merge branch 'rebasechunkpatches' of github.com:PaperMC/Paper into dev/chunksystem

Spottedleaf committed:
Rebuild patches

Spottedleaf committed:
Remove onChunkAdd from optimise anyPlayerCloseEnoughForSpawning patch

Handled by chunk system hooks

Spottedleaf committed:
Changes to MCUtil

- Make getChunkStatus use ChunkHolder#getChunkHolderStatus
- Make ensureMain use isMainThread

Spottedleaf committed:
Merge branch 'rebasechunkpatches' of github.com:PaperMC/Paper into dev/chunksystem

Spottedleaf committed:
Rebuild patches

Spottedleaf committed:
Fix import diff in DistanceManager

Spottedleaf committed:
Use new thread for async chunk debugs instead of asyncExecutor

The executor may be shutdown at this stage

Spottedleaf committed:
Return early from async debug chunks

Otherwise it will double run the command

Spottedleaf committed:
Merge branch 'master' of github.com:PaperMC/Paper into dev/chunksystem

Spottedleaf committed:
Merge branch 'master' of github.com:PaperMC/Paper into dev/chunksystem

Spottedleaf committed:
Rebuild patches

Spottedleaf committed:
Merge branch 'master' of github.com:PaperMC/Paper into dev/chunksystem

Spottedleaf committed:
Rebuild patches

Spottedleaf committed:
Merge branch 'master' of github.com:PaperMC/Paper into dev/chunksystem

Spottedleaf committed:
Rebuild patches

Spottedleaf committed:
Set unload cooldown to 10s explicitly

One of the earlier diffs to drop the api for explicit removeDelay
did not adjust the unload_cooldown timeout, which resulted
in the ticket never expiring resulting in a memory leak

Spottedleaf committed:
Completely block recursive chunk unloads

Stage 1 & 3 were blocked, but not stage 2.

Spottedleaf committed:
Prevent adding removed chunkholders into unload queue

Delayed scheduling of tasks may call checkUnload, which
would have placed the unloaded chunkholder back into queue.
For most callbacks, double unloading is not an issue, but
for the regionmanagers it is a significant issue.

Spottedleaf committed:
Adjust handling of mustNotSave and properly handle isUnsaved

The new isUnsaved handling also ensures we do not save
unloading chunks even if they do not need to be saved,
as was the behavior of the previous system.

The new handling of mustNotSave is to allow the entity chunk
to save, but not the poi chunk or the level chunk. While
this behavior is ridiculous, it was also (minus the poi part)
the behavior of the old chunk system which nobody complained about,
so I'm not going to bother to fix it.

Spottedleaf committed:
Ensure regionfile threads are saturated during flush saves

Fix the backoff timings in waitUntilAllExecuted, as they
were too large. Also, add a new method to RegionFileIOThread
that allows a partial flush. This partial flush is used
instead of the full flush so that we can ensure the
regionfile threads never become idle, but at the same
time ensure the queue size on the regionfile threads
does not exceed a small number of chunks so that
we do not run into memory issues during shutdown.

Spottedleaf committed:
Fix mustNotSave for chunk unloading

Can't pass the chunk or poichunk to the next unload stage

Spottedleaf committed:
Merge branch 'master' of github.com:PaperMC/Paper into dev/chunksystem

Spottedleaf committed:
Rebuild patches

Spottedleaf committed:
Do not process ticket levels off-main

While the chunk system is capable of performing ticket levels
asynchronously, the chunkholder create/delete/add/remove callbacks
are not safe to run asynchronously.

Spottedleaf committed:
Merge branch 'master' of github.com:PaperMC/Paper into dev/chunksystem

Spottedleaf committed:
Rebuild patches

Spottedleaf committed:
Make sure chunks unloaded with mustNotSave unload

By dropping the data from stage 1, we did not run
the unload logic in stage 2. Now, the data is simply
not saved in stage 2, but still unloaded.

Spottedleaf committed:
Merge branch 'master' of github.com:PaperMC/Paper into dev/chunksystem

Spottedleaf committed:
Rebuild patches

Spottedleaf committed:
Re-order starlight and dataconverter

Spottedleaf committed:
Revert optimise notify

Does not work with new chunk system, maybe will fix later

Spottedleaf committed:
Rebase fixups

Spottedleaf committed:
Move Rewrite Chunk system into first set of patches

This also removes the other older chunk system patches

Spottedleaf reviewed (commented):
Raw diff looks OK to me

Spottedleaf committed:
Add legacy chunk system patches to removed folder
